[NVFuser] Opinfos for extremal values

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76768
* __->__ #75917

Added slow tests for comparing the eager & fused outputs for given extremal inputs.

davidberard98 committed:
[NVFuser] Opinfos for extremal values in binary ufuncs

Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values in binary ufuncs"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

eellison commented:
I think you can add a label `slow-tests` to trigger on your pr

davidberard98 commented:
@eellison I think right now it's ciflow/trunk, it was changed ~ 1 month ago

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

ngimel commented on test/test_jit_cuda_fuser.py:
```diff
@@ -4559,12 +4568,64 @@ def test_nvfuser_correctness(self, device, dtype, op):
+            trace(*clone_inputs((sample.input, *sample.args)), **sample.kwargs)
+
+            for extremal_sample in _get_extremal_samples(sample, dtype):
+                ref = op(*clone_inputs((extremal_sample.input, *extremal_sample.args)), **extremal_sample.kwargs)
```
you should catch errors here, many eager ops (like cholesky or other linalg functions) will error out. If eager function errors out, we shouldn't attempt to run traced function

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot rebase this

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

mruberry commented on test/test_jit_cuda_fuser.py:
```diff
@@ -4588,12 +4597,77 @@ def test_nvfuser_correctness(self, device, dtype, op):
+    @unittest.skipIf(not RUN_NVFUSER, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    @ops(op_db, allowed_dtypes=(torch.float16, torch.bfloat16, torch.float32,
```
What happens if nvFuser doesn't support the operator?

davidberard98 replied:
We'll still run the test, but the operator won't be fused by nvfuser (so it'll just run the usual eager implementation of the op)

mruberry replied:
That kinda seems excessive, right? Not the end of the world. In the future it'd be nice to know which operators are supported by nvFuser so we don't excessively test

davidberard98 replied:
Is there a good way to keep an up to date list of ops supported by NVFuser? I imagine that it would be pretty easy to forget to update the list when NVFuser adds new ops

mruberry replied:
Yeah that's a good concern -- I don't know offhand

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

jjsjann123 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14584,6 +14597,9 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                # lambda impl
                DecorateInfo(unittest.expectedFailure, "TestNormalizeOperators", "test_normalize_operator_exhaustive"),
                DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
+               # 76571
```
Thanks for tracking this.
Linking the issue #76571

jjsjann123 reviewed (approved):
LGTM

jjsjann123 commented on test/test_jit_cuda_fuser.py:
```diff
@@ -4588,12 +4597,76 @@ def test_nvfuser_correctness(self, device, dtype, op):
+                    # if eager errors out, then don't expect NVFuser to pass
+                    continue
+
+                with freeze_rng_state():
```
nitpick: we've seen in a few cases where the test fails where there's no fusion happening.
I think those are still legit failures and should be reported by the tests. Does it make sense to explicitly report them by checking if the traced function has `CudaFusionGroup`?

davidberard98 replied:
To clarify - the idea here is to skip more checks, by skipping situations that don't have a `CudaFusionGroup` - is that right?

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] Opinfos for extremal values"


Added slow tests for comparing the eager & fused outputs for given extremal inputs.

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge this please
